pkg-manager/resolve-dependencies/src/resolvePeers.ts

```diff
@@ -113,13 +113,23 @@ export function resolvePeers<T extends PartialResolvedPackage> (
     dependenciesByProjectId[id] = mapValues((nodeId) => pathsByNodeId[nodeId], directNodeIdsByAlias)
   }
   if (opts.dedupePeerDependents) {
-    const depPathsMap = deduplicateDepPaths(depPathsByPkgId, depGraph)
-    Object.values(depGraph).forEach((node) => {
-      node.children = mapValues((childDepPath) => depPathsMap[childDepPath] ?? childDepPath, node.children)
-    })
-    for (const { id } of opts.projects) {
-      dependenciesByProjectId[id] = mapValues((depPath) => depPathsMap[depPath] ?? depPath, dependenciesByProjectId[id])
+    let duplicates = Object.values(depPathsByPkgId).filter((item) => item.length > 1)
+
+    function deduplicationPass () {
+      const res = deduplicateDepPaths(duplicates, depGraph)
+      const { depPathsMap } = res
+      duplicates = res.remainingDuplicates
+
+      Object.values(depGraph).forEach((node) => {
+        node.children = mapValues((childDepPath) => depPathsMap[childDepPath] ?? childDepPath, node.children)
+      })
+      for (const { id } of opts.projects) {
+        dependenciesByProjectId[id] = mapValues((depPath) => depPathsMap[depPath] ?? depPath, dependenciesByProjectId[id])
+      }
```

Comment:
do we need to do this on each cycle? Can't we do it one time after the while cycle?

Reply:
You're correct. Moved it out.

I was also getting paranoid about the while loop. Added a maxPasses guard. In the worst case scenario, where every duplicated package is dependent on another duplicated package, the max number of passes that can do anything is same as `duplicates.length`.